### DISCUSSION TITLE
TLSRoute match ALPN

### DISCUSSION THREAD

**Discussion by dboslee** [kubernetes-sigs/gateway-api#1683] [Ideas]
:wave: Hey everyone, I work at [Teleport](https://github.com/gravitational) and have been looking at Gateway API to replace an internal project we built to handle routing many services through a single gateway.

Our specific use case requires TLS inspection to route based on the TLS SNI and ALPN extensions. 

It seems the TLSRoute currently only supports routing based on the SNI so I would like to start the discussion of adding something like the following to the TLSRoute spec:

```golang
type TLSRouteRule struct {
    Matches []TLSRouteMatch `json:"matches.omitempty"`
    BackendRefs []BackendRef `json:"backendRefs,omitempty"`
}

type TLSRouteMatch struct {
    ALPN []TLSApplicationProtocol `json:"alpn.omitempty"`
}

type TLSApplicationProtocol string
```

This would allow service owners to specify a list of ALPNs to match and route to a given set of BackendRefs.

If this is something the community is in favor of I would be happy to write up a more exhaustive GEP and get involved in this project.

**Comment by howardjohn** [comment]
Just curious, can you expand on why you match on ALPN?

**Reply by dboslee** [reply]
We have multiple protocols wrapped in TLS using a single port and route to different services based on the ALPN. There are some more specifics [here](https://goteleport.com/docs/architecture/tls-routing/).

**Reply by howardjohn** [reply]
Just curious, why not match on SNI?

**Reply by dboslee** [reply]
Sorry missed this reply, we do use SNI to handle multi tenancy but use ALPN to route to different services for each tenant.

**Comment by youngnick** [comment]
On a (very) quick review, this looks reasonable to me:
- ALPN is roughly equivalent to SNI in terms of metadata for routing in that you don't _have_ to terminate TLS to have it available, so it fits the TLSRoute use case there
- This is an additive change, so we don't need an API revision.

I'm low bandwidth at the moment (travelling), so I probably won't make the meeting next week either, but this sounds pretty reasonable to me, especially given that TLSRoute hasn't graduated yet.

**Comment by howardjohn** [comment]
Isn't ALPN a negotiation? If we are passing through the TLS handshake, we need to pick the destination before we can negotiate. So we are really not routing on the *selected* ALPN, but the offered ones? Is that accurate?

What happens if I had match for `a` and `b`, and I get a request with `alpn: [a,b]`? Do we prioritize it by order? Is this something supported by proxies?

**Reply by dboslee** [reply]
Yes this is accurate. And I agree with prioritizing by order as the [RFC](https://www.rfc-editor.org/rfc/rfc7301.html) states:
> "ProtocolNameList" contains the list of protocols advertised by the client, in descending order of preference.

So the first application protocol _should_ be the clients preference. However I am not certain if all proxies would enforce this the same way. Would it make sense to leave this up to the implementation?

**Comment by howardjohn** [comment]
Do we have some info on what proxies support this? I have only found Envoy personally

**Reply by dboslee** [reply]
I have used nginx and envoy to do alpn based routing. See the nginx [ssl_preread_module docs](http://nginx.org/en/docs/stream/ngx_stream_ssl_preread_module.html). I've also looked at traefik which seems to support it [here](https://doc.traefik.io/traefik/master/routing/routers/#alpn) although I've never used it personally.

If we're worried to few proxies support this, is there a good way for extending the API for implementation specific features?

**Reply by youngnick** [reply]
> If we're worried to few proxies support this, is there a good way for extending the API for implementation specific features?

For most other areas, the answer is "yes, totally". But because this will require adding some fields to TLSRoute, unfortunately, we're stuck. That said, as I said above, I'm definitely in favor of exploring this more.

**Comment by youngnick** [comment]
Any chance you could bring this to a community meeting @dboslee? I'll add it to the agenda for tomorrow, but as it's President's Day in the US, I expect a quieter meeting. I think if we can get some more thumbs-up here, then you're clear to go for at least an initial clarification GEP that includes the What and Why, but leaves aside the How.

(Edit: that is, Step 2 on https://gateway-api.sigs.k8s.io/geps/overview/)

**Reply by dboslee** [reply]
Sound good. I should be able to make the meeting today

**Comment by shaneutt** [comment]
NGinx and Envoy can do this, but it is still not common across downstream implementations. I'm generally in favor of the next step here being a GEP to suggest adding it as an extended and experimental feature and we can spend some time in that state to see what implementations end up doing with it. :+1:

**Comment by rikatz** [comment]
@rostislavbobo for TLSRoute promotion I think we should keep this off the API, but we maybe want to keep it on track for Gateway API 1.6, wdyt?

**Reply by rikatz** [reply]
or once TLSRoute GEP is merged, this can probably be proposed as an experimental API/GEP (tho this still needs a GEP)

**Reply by youngnick** [reply]
Yes, agreed, this can be a separate GEP that adds an extended, experimental feature to TLSRoute once it's graduated.

**Reply by wiktor-k** [reply]
It seems that GEP [has been recently merged](https://github.com/kubernetes-sigs/gateway-api/pull/4064). Rendered output: https://gateway-api.sigs.k8s.io/geps/gep-2643/ (Just noting for other readers).

**Reply by rostislavbobo** [reply]
Agree to track ALPN as part of GW API 1.6 to prioritize TLSRoute graduation for 1.5, thanks
